src/ring_theory/noetherian.lean

```diff
@@ -334,6 +344,38 @@ begin
   exact fg_ker_comp f₁ g₁ hf (fg_restrict_scalars g.ker hg hsur) hsur
 end
 
+/-- A finitely generated idempotent ideal is generated by an idempotent element -/
+lemma _root_.ideal.is_idempotent_elem_iff_of_fg {R : Type*} [comm_ring R] (I : ideal R)
+  (h : I.fg) :
+  is_idempotent_elem I ↔ ∃ e : R, is_idempotent_elem e ∧ I = R ∙ e :=
+begin
+  split,
+  { intro e,
+    obtain ⟨r, hr, hr'⟩ := submodule.exists_mem_and_smul_eq_self_of_fg_of_le_smul I I h
+      (by { rw [smul_eq_mul], exact e.ge }),
+    simp_rw smul_eq_mul at hr',
+    refine ⟨r, hr' r hr, antisymm _ ((submodule.span_singleton_le_iff_mem _ _).mpr hr)⟩,
+    intros x hx,
+    rw ← hr' x hx,
+    exact ideal.mem_span_singleton'.mpr ⟨_, mul_comm _ _⟩ },
+  { rintros ⟨e, he, rfl⟩,
+    simp [is_idempotent_elem, ideal.span_singleton_mul_span_singleton, he.eq] }
+end
+
+lemma is_idempotent_elem_iff_eq_bot_or_top {R : Type*} [comm_ring R] [is_domain R]
```

Comment:
Shouldn't this be called `ideal.is_idempotent_elem_iff_eq_bot_or_top`?

Maybe the easiest thing to do would be to create a small new `namespace ideal` section just below and place the two relevant lemmas there. Maybe something like:
```lean
end submodule

namespace ideal

variables {R : Type*} [comm_ring R] {I : ideal R} (hfg : I.fg)
include hfg

lemma is_idempotent_elem_iff_of_fg : ...

lemma is_idempotent_elem_iff_eq_bot_or_top [is_domain R] : ...

end ideal
```